Reproduce states by letting each component opt in on handling state recovery itself

elupus opened the pull request:
This allow things like media player to turn on, set source, set sound mode
in one state declaration

## Description:

This is an attempt to solve #5123 and allow scene component to reproduce states using available services. It will no longer only call one service to set state. It will let each component handle it's own state restore processes by opting in.

## Background
Using service schema:s to automatically detect callable services was rejected as not flexible enough, and an explicit implementation was selected

**Related issue (if applicable):** https://github.com/home-assistant/home-assistant.io/pull/8410

## Example entry for `configuration.yaml` (if applicable):
```yaml
scene:
  - name: playback_movie
    entities:
      media_player.bedroom:
        state: on
        sound_mode: Dummy Movie
        volume_level: 1
      media_player.living_room:
        state: off
        sound_mode: Dummy Music

  - name: playback_music
    entities:
      media_player.bedroom:
        state: on
        sound_mode: Dummy Music
        volume_level: 0.5
      media_player.living_room:
        state: on
        sound_mode: Dummy Music

media_player:
    platform: demo
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code does not interact with devices:
  - [X] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

elupus commented:
Can someone add a wip label? Didn't prefix it by WIP: since i thought labels where preferred.

elupus commented:
One test failing still. Unsure how to fix however. The test calls turn_on on group domain. But have not mocked that. Seems the default turn_on service on group domain does not specify a schema?

MartinHjelmare commented:
I tried a similar logic to validate services from state here:
https://github.com/MartinHjelmare/home-assistant/blob/cddd07f1d892eaff34834817ad7232a2595a9d54/homeassistant/helpers/service.py#L115-L154

Some part of that logic could help here, eg finding only services with required attributes in schemas, as discussed in point 1 above.

My PR didn't make it, since it included core changes that wasn't the correct approach.

Maybe this PR can make it since it's more contained.

About discussion point 2, there's no `set_brightness` service for the light domain. Maybe that was just a bad example? Since we store the key representing the service in a dict, I would think that double calls wouldn't be a problem? Is it that we append the same entity_id to the list for the key? Then just make it a set instead of a list. Sets can only have unique values.

elupus commented:
whitelisting services with required parameters by default might work. Yeah your approach looks quite similar to mine. I think we can contain it to reproduce state logic. I will look through to see if I can copy some of your logic.

Do you happen to know where group.turn_on is implemented in tests and in standard? Seems it doesn't set schema in tests (not checked real version yet)

MartinHjelmare commented:
Here's the hass main component service registrations for turn on etc that are used for groups.

https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/__init__.py

elupus commented:
Hmm. Okey so there is no schema specified for groups. I would assume it still passes on attributes to entities in a group?

It's allowed to call group.turn_on with brightness attribute set?

Above complicates things. I suppose not supporting that use case could be a known limitation of the state component. But I'll see if I can semi handle it.

MartinHjelmare commented:
I tried to solve that here by expanding groups to individual entities first.
https://github.com/home-assistant/home-assistant/pull/8819

But we wanted to move the new function to the service helper module, and then we hit cyclic imports. The state helper is a mess.

elupus commented:
Right. Yes i was thinking of a per component state reproduction aswell. I just wonder of this lightweight change might be okey in the interm. I think a fallback for domains without schema (groups) that works according to old logic could easily be added. 

As a future step we could check if domain support a reproduce state service and only call that if it exist. That way we could handle the change gradually.

MartinHjelmare commented:
It would probably be good to ask for an opinion from core members either here or in discord, before spending too much time on a direction. We don't want more closed PRs like mine.

If you haven't already read it, here's more discussion on neighboring topic from my other closed PR:
https://github.com/home-assistant/home-assistant/pull/8851

I think it's great that you are working on solving this. :+1:

balloob commented:
So I have some thoughts about this. I never wrote them down before because the last thing I wanted is more work on my plate (and that's what usually happens when I write down my thoughts).

Anyway, I will write them down.

The current way scenes work in Home Assistant is shit (I wrote it). The reason it's bad is because we violated our own very design principle: no cross component things. So instead of having 1 helper that knows about all components and cannot be extended by custom components, we should do the opposite. Each component should get a new method:

```python
@bind_hass
async def async_reproduce_states(hass, states):
```

Then it's up to each component to be smart about it. Each component will end up with maybe a 10 line functions that is easy to manage, easy to test. The helper can disappear or be limited to forward the states to the right domains:

```python
to_call = {}

for state in states:
    to_call.set_default(state.domain, []).append(state)

if to_call:
    await asyncio.wait([
        getattr(hass.components, domain).async_reproduce_states(states)
        for domain, states in to_call.items()
    ])
``` 

On a second note about scenes. I have allowed the scene component to get other platforms besides Home Assistant, putting two complete different things in 1 component. We should split all but the Home Assistant scenes into a `scene_external` component. Then we can also easily start creating scenes while we're running etc.

elupus commented:
Sounds lika a good approach.. Will you accept a partial implementation from
me? Ie let's say light and media player, with fallback to old method for
other domains. Not sure I'm up to doing it across the board.

Den mån 26 nov. 2018 20:46 skrev Paulus Schoutsen <notifications@github.com
>:

> So I have some thoughts about this. I never wrote them down before because
> the last thing I wanted is more work on my plate (and that's what usually
> happens when I write down my thoughts).
>
> Anyway, I will write them down.
>
> The current way scenes work in Home Assistant is shit (I wrote it). The
> reason it's bad is because we violated our own very design principle: no
> cross component things. So instead of having 1 helper that knows about all
> components and cannot be extended by custom components, we should do the
> opposite. Each component should get a new method:
>
> @bind_hassasync def async_reproduce_states(hass, states):
>
> Then it's up to each component to be smart about it. Each component will
> end up with maybe a 10 line functions that is easy to manage, easy to test.
> The helper can disappear or be limited to forward the states to the right
> domains:
>
> to_call = {}
> for state in states:
>     to_call.set_default(state.domain, []).append(state)
> if to_call:
>     await asyncio.wait([
>         getattr(hass.components, domain).async_reproduce_states(states)
>         for domain, states in to_call.items()
>     ])
>
> On a second note about scenes. I have allowed the scene component to get
> other platforms besides Home Assistant, putting two complete different
> things in 1 component. We should split all but the Home Assistant scenes
> into a scene_external component. Then we can also easily start creating
> scenes while we're running etc.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/18700#issuecomment-441772681>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAdeoEsbfeQU86fvUOzkkZZupyia2T_iks5uzEUAgaJpZM4Yx-Zg>
> .
>

balloob commented:
Yeah that makes sense. We should not try to do everything at once. We can check if the component supports restoring state with `hasattr(getattr(hass.components, domain), 'async_restore_states')`

elupus commented:
So i started playing with this. Its somewhat hard to keep the domain implementation short (given its going to be duplicated per domain). Seems i'll end up with pretty much converting the old code (or my new code in this pull) to helpers called from the component. Similar to below.

It does drop the reverse import from helper, adding it from other side with the lookup tables stored in the components instead. Light turn on will likely need the schema based approach from this pull. So the helper as now defined wont work.

I think it feels right, but what do you guys think? 

```patch
diff --git a/homeassistant/components/media_player/__init__.py b/homeassistant/components/media_player/__init__.py
index 8530a01..72aaba5 100644
--- a/homeassistant/components/media_player/__init__.py
+++ b/homeassistant/components/media_player/__init__.py
@@ -12,6 +12,7 @@ import functools as ft
 import hashlib
 import logging
 from random import SystemRandom
+from typing import Iterable
 from urllib.parse import urlparse
 
 from aiohttp import web
@@ -27,14 +28,18 @@ from homeassistant.const import (
     SERVICE_MEDIA_SEEK, SERVICE_MEDIA_STOP, SERVICE_SHUFFLE_SET,
     SERVICE_TOGGLE, SERVICE_TURN_OFF, SERVICE_TURN_ON, SERVICE_VOLUME_DOWN,
     SERVICE_VOLUME_MUTE, SERVICE_VOLUME_SET, SERVICE_VOLUME_UP, STATE_IDLE,
-    STATE_OFF, STATE_PLAYING, STATE_UNKNOWN)
+    STATE_ON, STATE_OFF, STATE_PAUSED, STATE_PLAYING, STATE_UNKNOWN)
+from homeassistant.core import State
 from homeassistant.helpers.aiohttp_client import async_get_clientsession
 import homeassistant.helpers.config_validation as cv
 from homeassistant.helpers.config_validation import PLATFORM_SCHEMA  # noqa
 from homeassistant.helpers.entity import Entity
 from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.state import async_map_state_to_services
+from homeassistant.helpers.typing import HomeAssistantType
 from homeassistant.loader import bind_hass
 
+
 _LOGGER = logging.getLogger(__name__)
 _RND = SystemRandom()
 
@@ -178,6 +183,22 @@ ATTR_TO_PROPERTY = [
     ATTR_MEDIA_SHUFFLE,
 ]
 
+STATE_TO_SERVICE = {
+    STATE_ON: SERVICE_TURN_ON,
+    STATE_OFF: SERVICE_TURN_OFF,
+    STATE_PLAYING: SERVICE_MEDIA_PLAY,
+    STATE_IDLE: SERVICE_MEDIA_STOP,
+    STATE_PAUSED: SERVICE_MEDIA_PAUSE,
+}
+
+ATTRIBUTE_TO_SERVICE = {
+    ATTR_MEDIA_VOLUME_LEVEL: SERVICE_VOLUME_SET,
+    ATTR_MEDIA_VOLUME_MUTED: SERVICE_VOLUME_MUTE,
+    ATTR_MEDIA_POSITION: SERVICE_MEDIA_SEEK,
+    ATTR_SOUND_MODE: SERVICE_SELECT_SOUND_MODE,
+    ATTR_INPUT_SOURCE: SERVICE_SELECT_SOURCE,
+}
+
 
 @bind_hass
 def is_on(hass, entity_id=None):
@@ -191,6 +212,27 @@ def is_on(hass, entity_id=None):
                for entity_id in entity_ids)
 
 
+@bind_hass
+async def async_reproduce_states(hass: HomeAssistantType,
+                                 states: Iterable[State]) -> None:
+
+    calls = []
+
+    for state in states:
+        mapped = async_map_state_to_services(
+            hass, state, STATE_TO_SERVICE,
+            ATTRIBUTE_TO_SERVICE)
+
+        calls_in_sync = [
+            hass.services.async_call(DOMAIN, key, value, blocking=True)
+            for key, value in mapped.items()
+        ]
+
+        calls.append(asyncio.wait(calls_in_sync))
+
+    await asyncio.gather(*calls)
+
+
 WS_TYPE_MEDIA_PLAYER_THUMBNAIL = 'media_player_thumbnail'
 SCHEMA_WEBSOCKET_GET_THUMBNAIL = \
     websocket_api.BASE_COMMAND_MESSAGE_SCHEMA.extend({
diff --git a/homeassistant/helpers/state.py b/homeassistant/helpers/state.py
index 61dc49c..878b57d 100644
--- a/homeassistant/helpers/state.py
+++ b/homeassistant/helpers/state.py
@@ -3,17 +3,13 @@ import asyncio
 import datetime as dt
 import json
 import logging
-from collections import defaultdict
+from collections import defaultdict, OrderedDict
 from types import TracebackType
 from typing import (  # noqa: F401 pylint: disable=unused-import
     Awaitable, Dict, Iterable, List, Optional, Tuple, Type, Union)
 
 from homeassistant.loader import bind_hass
 import homeassistant.util.dt as dt_util
-from homeassistant.components.media_player import (
-    ATTR_MEDIA_CONTENT_ID, ATTR_MEDIA_CONTENT_TYPE, ATTR_MEDIA_SEEK_POSITION,
-    ATTR_MEDIA_VOLUME_LEVEL, ATTR_MEDIA_VOLUME_MUTED, SERVICE_PLAY_MEDIA,
-    SERVICE_SELECT_SOURCE, ATTR_INPUT_SOURCE)
 from homeassistant.components.notify import (
     ATTR_MESSAGE, SERVICE_NOTIFY)
 from homeassistant.components.sun import (
@@ -35,14 +31,13 @@ from homeassistant.components.cover import (
 from homeassistant.const import (
     ATTR_ENTITY_ID, ATTR_OPTION, ATTR_TEMPERATURE, SERVICE_ALARM_ARM_AWAY,
     SERVICE_ALARM_ARM_HOME, SERVICE_ALARM_DISARM, SERVICE_ALARM_TRIGGER,
-    SERVICE_LOCK, SERVICE_MEDIA_PAUSE, SERVICE_MEDIA_PLAY, SERVICE_MEDIA_STOP,
-    SERVICE_MEDIA_SEEK, SERVICE_TURN_OFF, SERVICE_TURN_ON, SERVICE_UNLOCK,
-    SERVICE_VOLUME_MUTE, SERVICE_VOLUME_SET, SERVICE_OPEN_COVER,
+    SERVICE_LOCK, SERVICE_TURN_OFF, SERVICE_TURN_ON, SERVICE_UNLOCK,
+    SERVICE_OPEN_COVER,
     SERVICE_CLOSE_COVER, SERVICE_SET_COVER_POSITION,
     SERVICE_SET_COVER_TILT_POSITION, STATE_ALARM_ARMED_AWAY,
     STATE_ALARM_ARMED_HOME, STATE_ALARM_DISARMED, STATE_ALARM_TRIGGERED,
     STATE_CLOSED, STATE_HOME, STATE_LOCKED, STATE_NOT_HOME, STATE_OFF,
-    STATE_ON, STATE_OPEN, STATE_PAUSED, STATE_PLAYING, STATE_UNKNOWN,
+    STATE_ON, STATE_OPEN, STATE_UNKNOWN,
     STATE_UNLOCKED, SERVICE_SELECT_OPTION)
 from homeassistant.core import State, DOMAIN as HASS_DOMAIN
 from homeassistant.util.async_ import run_coroutine_threadsafe
@@ -55,10 +50,6 @@ GROUP_DOMAIN = 'group'
 # Update this dict of lists when new services are added to HA.
 # Each item is a service with a list of required attributes.
 SERVICE_ATTRIBUTES = {
-    SERVICE_PLAY_MEDIA: [ATTR_MEDIA_CONTENT_TYPE, ATTR_MEDIA_CONTENT_ID],
-    SERVICE_MEDIA_SEEK: [ATTR_MEDIA_SEEK_POSITION],
-    SERVICE_VOLUME_MUTE: [ATTR_MEDIA_VOLUME_MUTED],
-    SERVICE_VOLUME_SET: [ATTR_MEDIA_VOLUME_LEVEL],
     SERVICE_NOTIFY: [ATTR_MESSAGE],
     SERVICE_SET_AWAY_MODE: [ATTR_AWAY_MODE],
     SERVICE_SET_FAN_MODE: [ATTR_FAN_MODE],
@@ -70,7 +61,6 @@ SERVICE_ATTRIBUTES = {
     SERVICE_SET_HOLD_MODE: [ATTR_HOLD_MODE],
     SERVICE_SET_OPERATION_MODE: [ATTR_OPERATION_MODE],
     SERVICE_SET_AUX_HEAT: [ATTR_AUX_HEAT],
-    SERVICE_SELECT_SOURCE: [ATTR_INPUT_SOURCE],
     SERVICE_SEND_IR_CODE: [ATTR_IR_CODE],
     SERVICE_SELECT_OPTION: [ATTR_OPTION],
     SERVICE_SET_COVER_POSITION: [ATTR_POSITION],
@@ -82,9 +72,6 @@ SERVICE_ATTRIBUTES = {
 SERVICE_TO_STATE = {
     SERVICE_TURN_ON: STATE_ON,
     SERVICE_TURN_OFF: STATE_OFF,
-    SERVICE_MEDIA_PLAY: STATE_PLAYING,
-    SERVICE_MEDIA_PAUSE: STATE_PAUSED,
-    SERVICE_MEDIA_STOP: STATE_IDLE,
     SERVICE_ALARM_ARM_AWAY: STATE_ALARM_ARMED_AWAY,
     SERVICE_ALARM_ARM_HOME: STATE_ALARM_ARMED_HOME,
     SERVICE_ALARM_DISARM: STATE_ALARM_DISARMED,
@@ -132,6 +119,31 @@ def get_changed_since(states: Iterable[State],
             if state.last_updated >= utc_point_in_time]
 
 
+def async_map_state_to_services(hass: HomeAssistantType,
+                                state: State,
+                                state_to_service: Dict[str, str],
+                                attribute_to_service: Dict[str, str]
+                                ) -> Dict[str, Dict]:
+    """Maps state value to a set of services to call"""
+    calls = OrderedDict()
+
+    if state.state in state_to_service:
+        calls[state_to_service[state.state]] = \
+            {ATTR_ENTITY_ID: state.entity_id}
+    elif state.state is None:
+        pass
+    else:
+        raise Exception("Unknown state type {}".format(state.state))
+
+    for key, value in state.attributes.items():
+        if key in attribute_to_service:
+            calls[attribute_to_service[key]] = \
+                {ATTR_ENTITY_ID: state.entity_id,
+                 key: value}
+    _LOGGER.debug(calls)
+    return calls
+
+
 @bind_hass
 def reproduce_state(hass: HomeAssistantType,
                     states: Union[State, Iterable[State]],
@@ -145,6 +157,26 @@ def reproduce_state(hass: HomeAssistantType,
 async def async_reproduce_state(hass: HomeAssistantType,
                                 states: Union[State, Iterable[State]],
                                 blocking: bool = False) -> None:
+
+    if isinstance(states, State):
+        states = [states]
+
+    to_call = defaultdict(list)
+
+    for state in states:
+        to_call[state.domain].append(state)
+
+    if to_call:
+        await asyncio.gather(*[
+            getattr(hass.components, domain).async_reproduce_states(states)
+            for domain, states in to_call.items()
+        ])
+
+
+@bind_hass
+async def async_reproduce_state_old(hass: HomeAssistantType,
+                                    states: Union[State, Iterable[State]],
+                                    blocking: bool = False) -> None:
     """Reproduce given state."""
     if isinstance(states, State):
         states = [states]

```

balloob commented:
Something like that would work. I would suggest that we move all the state restore of media player to a `media_player/_restore_state.py` and just import `async_reproduce_states` into the media player component. That way we can keep related code together for a component.

MartinHjelmare commented:
For consistency, I think all components should use voluptuous schemas to create valid service calls from states.

We can use the core class `ServiceCall` to store the info we need for each service call.

We should probably keep the `context` of the state when creating the service call. The state context will be created by the user when configuring the scene.

elupus commented:
Those two comments in combination get's somewhat messy. Since i need to get schema which currently is stored in __init__.py. So i'd need to cross import the two files, or move all schemas to separate file. I suppose _restore_state.py could get the service schema:s from the registered services like i do in this pull request instead of internal of the component, but that seem wrong.

MartinHjelmare commented:
We could eg have one sub module called `_services.py` where we define service related things like service schemas. Then we have another module called `_states.py` where states for the component are defined and how to reproduce those states.

We don't import things from the base component to these submodules.

elupus commented:
Before i fixup tests. is it okey to drop the grouping of same service calls. Seems of limited use, i assume its performance related.

MartinHjelmare commented:
> Some "annoyances". climate.ecobee and switch.mysensors which i can't import from the base climate/init.py or switch/init.py.
>
> It would not help to move it into _state_restore.py, since same cyclic import would occur. I'd have to split the schema from those two platforms to get that to work. Should we forward the state restore into the platform implementation component as well in some way?

@balloob has said earlier in my PR that we don't need to support reproduce states for platforms, so in the longer run that will be a breaking change. Until there's a nice way to support platforms, I agree with that.

MartinHjelmare commented on homeassistant/components/climate/__init__.py:
```diff
@@ -135,6 +143,35 @@
+
+}
+
+SERVICE_TO_SCHEMA = {
```
Service schemas are already available in the service registry in the core, and can be accessed like `hass.services.async_services`.

elupus replied:
Ok. Just thought it felt wrong to involve global stuff in the component when the component knows self.

MartinHjelmare replied:
It's the component that has registered its services on the core. We shouldn't duplicate that info.

MartinHjelmare replied:
We should still parse this in the component, but we can do it simpler.
```py
service_schemas = {
    service: service_obj.schema
    for service: service_obj
    in hass.services.async_services()[domain]}
```

MartinHjelmare commented on homeassistant/helpers/state.py:
```diff
@@ -132,6 +99,103 @@ def get_changed_since(states: Iterable[State],
+
+        def add_if_complete(service: str) -> None:
+            """Add a service call with the attributes needed."""
+            schema = voluptuous.schema_builder.Schema(
```
It's more flexible if we define the schema in each component, and accept the schema in the helper.

All voluptuous names are available under `voluptuous.`, so `voluptuous.Schema` is enough. We usually import `voluptuous as vol`.

elupus replied:
It's like that already. It's just that i need to modify the schema to drop extra data and not bail out on extra attributes.

MartinHjelmare replied:
I mean that instead of passing a dict that maps states to services we pass a schema.

MartinHjelmare replied:
Another point is that it's more flexible if we keep the knowledge about how he schema is made in each component, and just let the helper validate the state against the schema and if it passes, store the service call.

elupus replied:
we still need states to service and whitelisted services. some services take no parameter or toggle states. for example light.toggle. it takes optional transition. This services can't be allowed (with default schema anyway).

Also if it is allowed with required transition attribute, its one of those that require consumption of attributes. Otherwise turn_on will turn on, with a transition for example, the the toggle service gets called with the remaining attribute.

MartinHjelmare replied:
I'm not saying we should use the default schema. I'm saying we should define each schema in each component. We can build off the service schema when available. If not available, the component makes one from scratch.

MartinHjelmare replied:
I'm don't understand the turn on + transition + toggle example. If we turn on, we don't toggle, right?

elupus replied:
I'm not sure i follow. We have two things in play:
1: we need to map "state: on/off/heat" to a service taking no parameters.
2: we need to map additional attributes to additional services.

For 1, i don't think we can do without a dict? they cant be differentiated by schema.
For 2. the schema is already passed from component.

MartinHjelmare replied:
A schema can use a dict to validate.

elupus replied:
We need to add the allowance of extra attributes somewhere. Otherwise the code ends up only able to call a single service as the original.

So the schemas registered for the service in the registry are never ever okey as is.

MartinHjelmare replied:
Yep, it's fine to make new schemas based on the existing service schemas. But the helper shouldn't do that. That's all I'm saying.

MartinHjelmare commented on homeassistant/helpers/state.py:
```diff
@@ -132,6 +99,103 @@ def get_changed_since(states: Iterable[State],
+            # for same attribute if schema would
+            # allow it
+            for key in attributes:
+                expanded_attributes.pop(key, None)
```
Is this a problem we've encountered or is it just theoretical? As far as I know, there are no shared attributes between services for a domain (for services that are not directly affecting a state, eg on/off).

I think we could just assume that we'll continue following that rule, and make it a rule when defining service attributes. Ie attribute names should be unique per domain for services that only modify state attributes.

elupus replied:
light.turn_on light.toggle share transition.

elupus replied:
climate.set_temperature climate.set_operation_mode share operation mode.

MartinHjelmare replied:
We will never call turn on and toggle at the same time, right?

It's weird that `set_temperature` takes operation mode. We should probably look at removing that for the future. But for now we can handle that in the climate component.

MartinHjelmare replied:
`set_temperature` also requires other attributes, so only operation_mode won't trigger both services.

elupus replied:
nope, but if you do set a temperature and give it operation mode, you get two calls.

elupus replied:
> We will never call turn on and toggle at the same time, right?
For state restore it would if we don't consume the transition when calling turn on. The schema for toggle would still match.

MartinHjelmare replied:
But the schema shouldn't both validate turn on and toggle. This is why we should make new schemas in each component.

MartinHjelmare replied:
It sounds good to combine state and state attributes into a single validation schema, since we don't want to call more than one service that changes state but we can allow many services that changes different attributes.

About your example, I'm not sure that `transition` without state should be allowed. The transition attribute is connected to a service that should change the state. So we should require the state in the scene.

MartinHjelmare commented on homeassistant/helpers/state.py:
```diff
@@ -132,6 +99,103 @@ def get_changed_since(states: Iterable[State],
+                            context=state.context))
+
+        # first try to reproce the state value, taking all attributes
+        if state.state in state_to_service:
```
The following code block might also be better to keep in each component. The priority between states and state attributes might be different for different components.

elupus replied:
sure, i'll split to two calls.

elupus commented:
I just pushed my current branch. None of the discussion points have been addressed there yet. Only fixed some tests and added alarm_control_panel as well.

elupus commented:
@MartinHjelmare i think this ended up neater with the extended schema. There is now a helper function for components to optionally use which creates extended schemas based on a list of services and extra attributes (state only used now). This helper makes use of registered schemas for services like requested. The component can create the list itself or it can adjust it if needed.

Then the component calls another helper function taking a list of states and a list of schema,service tuples and if the schema matches, calls the services.

This ended up quite neat. Not too much extra in component, and few spacial cases. I think i can get @balloob request of separate file working now with too much rework.

Ps. I'll rebase and squash a lot of the commits before done. History on branch is getting quite cluttered with different attempts.

MartinHjelmare commented on homeassistant/helpers/state.py:
```diff
@@ -93,88 +93,90 @@ def get_changed_since(states: Iterable[State],
+    """Helper function to construct an extended schema for state restore."""
+
+    for service, extension in services:
+        services = hass.services.async_services()[domain]
```
Looks like we're modifying the list we're iterating here.

elupus replied:
Indeed. Odd it worked. Will fix

MartinHjelmare commented on homeassistant/helpers/state.py:
```diff
@@ -132,6 +89,109 @@ def get_changed_since(states: Iterable[State],
+            # for same attribute if schema would
+            # allow it
+            for key in validated:
+                attributes.pop(key, None)
```
I'm still hesitant that we should handle this case in the helper. It makes the helper more unpredictable, regarding what service calls it will return.

elupus replied:
I cant allow both set operation mode and set temperature service for climate if not. Well i can allow it, but well get double calls.

I think it should be like this, but I can live with dropping it so I'll do that.

MartinHjelmare replied:
Can we handle the climate case in the climate component instead? Maybe use a custom validation schema?

elupus replied:
Well, the schema must also map to a valid service call. Ie all attributes needed for the service and no more.

But we could use a custom schema for the set temperature service skipping operation mode since that is optional in the service. But given the climate architecture redesign peneing. I'm not sure it's worth it.

balloob replied:
It's totally fine to call multiple services to recreate a single scene state.

elupus replied:
The issue is that we can end up calling multiple services for a single attribute. Should normally not cause a problem as long as both services assume the same attribute have the same meaning and it's not some sort of toggling feature.

MartinHjelmare commented on homeassistant/helpers/state.py:
```diff
@@ -145,76 +205,21 @@ def reproduce_state(hass: HomeAssistantType,
+
+    if to_call:
+        await asyncio.gather(*[
+            getattr(hass.components, domain).async_reproduce_states(
```
Do we need to handle the case of missing component for a domain?

elupus replied:
Yea. I was thinking of adding a fallback for the "standard" services turn_on/turn_off.

MartinHjelmare commented on homeassistant/helpers/state.py:
```diff
@@ -53,42 +32,20 @@
-    SERVICE_SET_COVER_TILT_POSITION: [ATTR_TILT_POSITION]
+# Each item is a service which has unique named attribute for
+# the domain.
+SERVICE_WHITELIST = {
```
Do we still need this?

elupus replied:
Nope. Just so I remember what components remain to do.

elupus commented on homeassistant/components/climate/__init__.py:
```diff
@@ -5,21 +5,27 @@
 
 import voluptuous as vol
 
+from homeassistant.const import (
```
Yes.. but other component need them from climate base. Can be fixed separate. I don't want to grow this pull even further. (this is due to move to _restore_state.py)

elupus commented:
I need some encouraging words @balloob , @MartinHjelmare. This pull has grown way too large from it's initial of touching pretty much one single file (with no more features nor change in behaviour).

The move to separate _restore_state.py caused quite some pain. climate, media_player needed moving stuff to core.const. group needs the function to expand groups in __init__.py so i have not moved that one. I'm very much unsure if there really was a benefit of moving it to _restore_state.py.

MartinHjelmare commented:
Yeah, it doesn't look clean, eg we need to redefine the domain constant in `_reproduce_state.py.

We shouldn't call the module restore_state either. That's another feature in home assistant, about persistent state between restarts.

@balloob what do you say?

balloob commented:
So I haven't really followed this PR too much.

From looking at the PR, I am surprised how it was implemented. I thought that it would be opt-in for components, falling back to the original method if it didn't exist. That way only 1 component would have to be migrated to the new logic and it would help in hashing out how it will work.

I do see now that moving to a separate `_reproduce_state.py` has lead to some circular dependencies. We should not move the states to `homeassistant.const`. We should consider either importing the necessary states inside the functions inside `_reproduce_state` or we don't extract it into a file.

balloob commented:
Yeah, restore_state is not a good name, reproduce_state is better, as that's what we're aiming to do.

elupus commented:
> I don't think you understand what I'm trying to say. I was suggesting we **only** implement this for 1 component, for example the switch component, because it's so simple. Forget about alarm control panel, forget about light, etc etc. Have the restore state look if a component supports this, otherwise fallback to old logic.
> 
> That way you're changing 1 component, not 20.

I do get what you are saying. Easier to handle any regressions too if only one part is changing.

I still need the helpers functions somewhere. So I'd be duplicating logic, and it needs to hook into the current code which isn't too easy given the way it groups up data. I'll see if i can hook it up without dropping the old logic. I'll probably update the pull request with current state. Squashed it together so it's less commits.

elupus committed:
Move group to it's own setup

elupus commented:
Okey.. so here is a slimmed down version. I've reworked and split the commit series so i can add them one by one (ie i have the other components locally moved), but here i've only added media_player and group. The old logic remains for other services.

balloob commented on homeassistant/components/media_player/_reproduce_state.py:
```diff
@@ -0,0 +1,45 @@
+        (SERVICE_PLAY_MEDIA, {}),
+    ]
+
+    service_schemas = async_construct_service_schemas(
```
I am not a big fan of extracting it from the voluptuous schemas. Why can't we do something like:

```
if VOLUME in state.attributes:
    await hass.services.async_call(DOMAIN, SERVICE_VOLUME_SET, blocking=True, context=context)

# etc
```

Because I wouldn't want people to start limiting their architecture to have to support this abstraction.

elupus replied:
Works reasonable for services with single attributes. Gets messy for ones with multiple optional attributes like lights turn_on, set_temperature. You need extract the optional attributes somehow.

I can see the benefits of explicit calls, but it will grow codebase.

elupus replied:
```patch
diff --git a/homeassistant/components/media_player/_reproduce_state.py b/homeassistant/components/media_player/_reproduce_state.py
index 6f053c4..ab7acd4 100644
--- a/homeassistant/components/media_player/_reproduce_state.py
+++ b/homeassistant/components/media_player/_reproduce_state.py
@@ -1,45 +1,126 @@
 """Module that groups code required to handle state restore for component."""
+import asyncio
 from typing import Iterable
 
 from homeassistant.const import (
-    ATTR_STATE, SERVICE_MEDIA_PAUSE, SERVICE_MEDIA_PLAY, SERVICE_MEDIA_SEEK,
+    SERVICE_MEDIA_PAUSE, SERVICE_MEDIA_PLAY, SERVICE_MEDIA_SEEK,
     SERVICE_MEDIA_STOP, SERVICE_TURN_OFF, SERVICE_TURN_ON, SERVICE_VOLUME_MUTE,
     SERVICE_VOLUME_SET, STATE_IDLE, STATE_OFF, STATE_ON, STATE_PAUSED,
     STATE_PLAYING)
 from homeassistant.core import State
-from homeassistant.helpers.state import (
-    async_construct_service_schemas, async_reproduce_domain_state)
 from homeassistant.helpers.typing import HomeAssistantType
 from homeassistant.loader import bind_hass
 
 
-@bind_hass
-async def async_reproduce_states(hass: HomeAssistantType,
-                                 states: Iterable[State]) -> None:
+def generate_data_dict(entity_id, attributes = {}, keys = ()):
+    data = {}
+    data['entity_id'] = entity_id
+    for key in keys:
+        if key in attributes:
+            data[key] = attributes[key]
+    return data
+
+
+async def async_reproduce_state(hass: HomeAssistantType,
+                                state: State) -> None:
     """Reproduce component states."""
     from . import (
+        ATTR_MEDIA_VOLUME_LEVEL,
+        ATTR_MEDIA_VOLUME_MUTED,
+        ATTR_MEDIA_SEEK_POSITION,
+        ATTR_INPUT_SOURCE,
+        ATTR_SOUND_MODE,
+        ATTR_MEDIA_CONTENT_TYPE,
+        ATTR_MEDIA_CONTENT_ID,
+        ATTR_MEDIA_ENQUEUE,
         SERVICE_PLAY_MEDIA,
         SERVICE_SELECT_SOURCE,
         SERVICE_SELECT_SOUND_MODE,
         DOMAIN,
     )
 
-    SERVICE_STATE_LIST = [
-        (SERVICE_TURN_ON, {ATTR_STATE: STATE_ON}),
-        (SERVICE_TURN_OFF, {ATTR_STATE: STATE_OFF}),
-        (SERVICE_MEDIA_PLAY, {ATTR_STATE: STATE_PLAYING}),
-        (SERVICE_MEDIA_STOP, {ATTR_STATE: STATE_IDLE}),
-        (SERVICE_MEDIA_PAUSE, {ATTR_STATE: STATE_PAUSED}),
-        (SERVICE_VOLUME_SET, {}),
-        (SERVICE_VOLUME_MUTE, {}),
-        (SERVICE_MEDIA_SEEK, {}),
-        (SERVICE_SELECT_SOURCE, {}),
-        (SERVICE_SELECT_SOUND_MODE, {}),
-        (SERVICE_PLAY_MEDIA, {}),
-    ]
-
-    service_schemas = async_construct_service_schemas(
-        hass, DOMAIN, SERVICE_STATE_LIST)
-
-    await async_reproduce_domain_state(
-        hass, DOMAIN, states, service_schemas)
+    if state.state == STATE_ON:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_TURN_ON,
+            generate_data_dict(state.entity_id),
+            blocking=True, context=state.context)
+    elif state.state == STATE_OFF:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_TURN_OFF,
+            generate_data_dict(state.entity_id),
+            blocking=True, context=state.context)
+    elif state.state == STATE_PLAYING:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_MEDIA_PLAY,
+            generate_data_dict(state.entity_id),
+            blocking=True, context=state.context)
+    elif state.state == STATE_IDLE:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_MEDIA_STOP,
+            generate_data_dict(state.entity_id),
+            blocking=True, context=state.context)
+    elif state.state == STATE_PAUSED:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_MEDIA_PAUSE,
+            generate_data_dict(state.entity_id),
+            blocking=True, context=state.context)
+
+    if ATTR_MEDIA_VOLUME_LEVEL in state.attributes:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_VOLUME_SET,
+            generate_data_dict(state.entity_id,
+                               state.attributes,
+                               (ATTR_MEDIA_VOLUME_LEVEL)),
+            blocking=True, context=state.context)
+
+    if ATTR_MEDIA_VOLUME_MUTED in state.attributes:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_VOLUME_MUTE,
+            generate_data_dict(state.entity_id,
+                               state.attributes,
+                               (ATTR_MEDIA_VOLUME_MUTED)),
+            blocking=True, context=state.context)
+
+    if ATTR_MEDIA_SEEK_POSITION in state.attributes:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_MEDIA_SEEK,
+            generate_data_dict(state.entity_id,
+                               state.attributes,
+                               (ATTR_MEDIA_SEEK_POSITION)),
+            blocking=True, context=state.context)
+
+    if ATTR_INPUT_SOURCE in state.attributes:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_SELECT_SOURCE,
+            generate_data_dict(state.entity_id,
+                               state.attributes,
+                               (ATTR_INPUT_SOURCE)),
+            blocking=True, context=state.context)
+
+    if ATTR_SOUND_MODE in state.attributes:
+        await hass.services.async_call(
+            DOMAIN, SERVICE_SELECT_SOUND_MODE,
+            generate_data_dict(state.entity_id,
+                               state.attributes,
+                               (ATTR_SOUND_MODE)),
+            blocking=True, context=state.context)
+
+    if (ATTR_MEDIA_CONTENT_TYPE in state.attributes) and \
+       (ATTR_MEDIA_CONTENT_ID in state.attributes):
+        await hass.services.async_call(
+            DOMAIN, SERVICE_PLAY_MEDIA,
+            generate_data_dict(state.entity_id,
+                               state.attributes,
+                               (ATTR_MEDIA_CONTENT_TYPE,
+                                ATTR_MEDIA_CONTENT_ID,
+                                ATTR_MEDIA_ENQUEUE)),
+            blocking=True, context=state.context)
+
+
+@bind_hass
+async def async_reproduce_states(hass: HomeAssistantType,
+                                 states: Iterable[State]) -> None:
+
+    await asyncio.gather(*[
+        async_reproduce_state(hass, state)
+        for state in states])
```


This would be the equivalent explicit version. It's much easier for a typo to sneak in due to the amount of duplicated code. Easy to miss to extend should an attribute be added to a service.

elupus replied:
@balloob do you like this approach better? If so i'll update the pull request. We can tackle the cyclic imports once the architecture discussion has come to fruition. It's not too bad to import in the function for now imho.

balloob replied:
I like this approach better.

balloob commented on homeassistant/components/media_player/_reproduce_state.py:
```diff
@@ -0,0 +1,45 @@
+
+
+@bind_hass
+async def async_reproduce_states(hass: HomeAssistantType,
```
We need to pass context so that state changes are attributes to the right initiator.

elupus replied:
Context is part of the State object already.

balloob replied:
context of the state object is the stored state, we need to pass the context of the one that invoked the action (either being a service call or automation/script)

elupus replied:
Btw. while i'm looking here. Do i need to support the blocking=false case?

balloob replied:
nah, we should drop that.

balloob commented:
We've done some thinking and we might have a solution for the circular imports: https://github.com/home-assistant/architecture/issues/124

balloob commented on homeassistant/components/media_player/_reproduce_state.py:
```diff
@@ -0,0 +1,45 @@
+"""Module that groups code required to handle state restore for component."""
```
Let's rename this file to `reproduce_state`.

balloob commented on homeassistant/components/media_player/_reproduce_state.py:
```diff
@@ -0,0 +1,45 @@
+async def async_reproduce_states(hass: HomeAssistantType,
+                                 states: Iterable[State]) -> None:
+    """Reproduce component states."""
+    from . import (
```
Let's move these constants to `media_player/const.py`

elupus replied:
Ie like the following to start with. So i don't have to update all other components potentially using the constants.

```patch
diff --git a/homeassistant/components/media_player/__init__.py b/homeassistant/components/media_player/__init__.py
index dbf5d9e..716d3de 100644
--- a/homeassistant/components/media_player/__init__.py
+++ b/homeassistant/components/media_player/__init__.py
@@ -35,12 +35,12 @@ from homeassistant.helpers.entity import Entity
 from homeassistant.helpers.entity_component import EntityComponent
 from homeassistant.loader import bind_hass
 
+from .const import *
 from .reproduce_state import async_reproduce_states  # noqa
 
 _LOGGER = logging.getLogger(__name__)
 _RND = SystemRandom()
 
-DOMAIN = 'media_player'
 DEPENDENCIES = ['http']
 
 ENTITY_ID_FORMAT = DOMAIN + '.{}'
@@ -56,38 +56,6 @@ ENTITY_IMAGE_CACHE = {
     CACHE_MAXSIZE: 16
 }
 
-SERVICE_PLAY_MEDIA = 'play_media'
-SERVICE_SELECT_SOURCE = 'select_source'
-SERVICE_SELECT_SOUND_MODE = 'select_sound_mode'
-SERVICE_CLEAR_PLAYLIST = 'clear_playlist'
-
-ATTR_MEDIA_VOLUME_LEVEL = 'volume_level'
-ATTR_MEDIA_VOLUME_MUTED = 'is_volume_muted'
-ATTR_MEDIA_SEEK_POSITION = 'seek_position'
-ATTR_MEDIA_CONTENT_ID = 'media_content_id'
-ATTR_MEDIA_CONTENT_TYPE = 'media_content_type'
-ATTR_MEDIA_DURATION = 'media_duration'
-ATTR_MEDIA_POSITION = 'media_position'
-ATTR_MEDIA_POSITION_UPDATED_AT = 'media_position_updated_at'
-ATTR_MEDIA_TITLE = 'media_title'
-ATTR_MEDIA_ARTIST = 'media_artist'
-ATTR_MEDIA_ALBUM_NAME = 'media_album_name'
-ATTR_MEDIA_ALBUM_ARTIST = 'media_album_artist'
-ATTR_MEDIA_TRACK = 'media_track'
-ATTR_MEDIA_SERIES_TITLE = 'media_series_title'
-ATTR_MEDIA_SEASON = 'media_season'
-ATTR_MEDIA_EPISODE = 'media_episode'
-ATTR_MEDIA_CHANNEL = 'media_channel'
-ATTR_MEDIA_PLAYLIST = 'media_playlist'
-ATTR_APP_ID = 'app_id'
-ATTR_APP_NAME = 'app_name'
-ATTR_INPUT_SOURCE = 'source'
-ATTR_INPUT_SOURCE_LIST = 'source_list'
-ATTR_SOUND_MODE = 'sound_mode'
-ATTR_SOUND_MODE_LIST = 'sound_mode_list'
-ATTR_MEDIA_ENQUEUE = 'enqueue'
-ATTR_MEDIA_SHUFFLE = 'shuffle'
-
 MEDIA_TYPE_MUSIC = 'music'
 MEDIA_TYPE_TVSHOW = 'tvshow'
 MEDIA_TYPE_MOVIE = 'movie'
diff --git a/homeassistant/components/media_player/reproduce_state.py b/homeassistant/components/media_player/reproduce_state.py
index 46352e3..5feafad 100644
--- a/homeassistant/components/media_player/reproduce_state.py
+++ b/homeassistant/components/media_player/reproduce_state.py
@@ -11,25 +11,26 @@ from homeassistant.core import Context, State
 from homeassistant.helpers.typing import HomeAssistantType
 from homeassistant.loader import bind_hass
 
+from .const import (
+    ATTR_MEDIA_VOLUME_LEVEL,
+    ATTR_MEDIA_VOLUME_MUTED,
+    ATTR_MEDIA_SEEK_POSITION,
+    ATTR_INPUT_SOURCE,
+    ATTR_SOUND_MODE,
+    ATTR_MEDIA_CONTENT_TYPE,
+    ATTR_MEDIA_CONTENT_ID,
+    ATTR_MEDIA_ENQUEUE,
+    SERVICE_PLAY_MEDIA,
+    SERVICE_SELECT_SOURCE,
+    SERVICE_SELECT_SOUND_MODE,
+    DOMAIN,
+)
+
 
 async def _async_reproduce_states(hass: HomeAssistantType,
                                   state: State,
                                   context: Optional[Context] = None) -> None:
     """Reproduce component states."""
-    from . import (
-        ATTR_MEDIA_VOLUME_LEVEL,
-        ATTR_MEDIA_VOLUME_MUTED,
-        ATTR_MEDIA_SEEK_POSITION,
-        ATTR_INPUT_SOURCE,
-        ATTR_SOUND_MODE,
-        ATTR_MEDIA_CONTENT_TYPE,
-        ATTR_MEDIA_CONTENT_ID,
-        ATTR_MEDIA_ENQUEUE,
-        SERVICE_PLAY_MEDIA,
-        SERVICE_SELECT_SOURCE,
-        SERVICE_SELECT_SOUND_MODE,
-        DOMAIN,
-    )
 
     async def call_service(service: str, keys=()):
         data = {}

```

elupus committed:
Let each component to handle restore of state

elupus commented:
Updated pull with new approach. climate, media_player and group done for now. Ive not moved the constants yet. Since that will affect other platforms, i'd like to avoid that in this pull (at least i need ok in current form first)

MartinHjelmare commented on homeassistant/components/climate/reproduce_state.py:
```diff
@@ -0,0 +1,105 @@
+        DOMAIN,
+    )
+
+    from .ecobee import (
```
Common component modules shouldn't be depend on platforms.

MartinHjelmare replied:
It's an architecture decision. We don't want to have that dependency. At least so far we haven't.

MartinHjelmare commented on homeassistant/components/group/reproduce_state.py:
```diff
@@ -0,0 +1,24 @@
+                                 context: Optional[Context] = None) -> None:
+    """Reproduce component states."""
+    from . import get_entity_ids
+    from homeassistant.helpers.state import async_reproduce_state
```
This can be moved to the top of the module.

MartinHjelmare replied:
Yeah, just this line.

elupus replied:
Nope. Not until switch has been converted to new system.

```
tests/helpers/test_state.py:12: in <module>
    from homeassistant.helpers import state
homeassistant/helpers/state.py:17: in <module>
    from homeassistant.components.switch.mysensors import (
homeassistant/components/switch/__init__.py:20: in <module>
    from homeassistant.components import group
homeassistant/components/group/__init__.py:26: in <module>
    from .reproduce_state import async_reproduce_states  # noqa
homeassistant/components/group/reproduce_state.py:5: in <module>
    from homeassistant.helpers.state import async_reproduce_state
E   ImportError: cannot import name 'async_reproduce_state'

```

MartinHjelmare replied:
Ok. :+1:

MartinHjelmare commented:
@balloob should confirm.

balloob commented:
We should make this a breaking change and drop Ecobee.

We will need tests added for the reproduce state functions for the components (that's why I commented earlier to just limit it to the switch component)

elupus committed:
Move constants for climate into const.py

For now import all into __init__.py to keep backword compat

elupus committed:
Move climate to it's own file

elupus committed:
Move media plyaer constants to const.py file

For now import all constants into __init__.py to keep
backword compatibility

elupus committed:
Move media player to it's own file

elupus committed:
Remove ecobee service from common components

BREAKING CHANGE

elupus committed:
Add tests for climate

elupus committed:
Add test for media_player

elupus commented:
Ah. crap start import wasn't allowed even in tests. Will fix. But apart from that (and missing tests for groups), do test look okey?

elupus committed:
Make sure we clone timestamps of state

elupus committed:
Add tests for groups

elupus committed:
Remove old tests for media player, it's handled by other tests

elupus committed:
Add tests for calls to component functions

elupus commented:
Rebased due to merge conflict. Will check what travis say about coverage of tests

elupus commented on homeassistant/components/climate/__init__.py:
```diff
@@ -21,25 +21,17 @@
     STATE_ON, STATE_OFF, TEMP_CELSIUS, PRECISION_WHOLE,
     PRECISION_TENTHS)
 
+from .const import *
```
Ok if i add noqa om this? I can provide a separate pull fixing this. But i rather not let this pull spread into specific component.

balloob replied:
I think that we should just fix it, it shouldn't be too difficult to just list the items?

elupus commented:
It will end up being ALL constants since components like ecobee and the lot
will import the constants from climate and not climate.const. also it needs
to import with # noqa for the same reason.

If we first fix components to import from const, the noqa issue goes away,
and list only ends up being what is used in __init__.py



Den mån 4 feb. 2019 19:35 skrev Paulus Schoutsen <notifications@github.com>:

> *@balloob* commented on this pull request.
> ------------------------------
>
> In homeassistant/components/climate/__init__.py
> <https://github.com/home-assistant/home-assistant/pull/18700#discussion_r253588845>
> :
>
> > @@ -21,25 +21,17 @@
>      STATE_ON, STATE_OFF, TEMP_CELSIUS, PRECISION_WHOLE,
>      PRECISION_TENTHS)
>
> +from .const import *
>
> I think that we should just fix it, it shouldn't be too difficult to just
> list the items?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/18700#discussion_r253588845>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAdeoD_s2Vl5WgqkO5w89Gix-rfIbIxwks5vKHzrgaJpZM4Yx-Zg>
> .
>

balloob commented:
The problem is that `import *` breaks all the linting checks for those files, we don't want that. You should be able to search & replace all in the code.

- Search `from homeassistant.components.climate import`
- Replace `from homeassistant.components.climate.const import`

elupus commented:
There are about 160 files affected for climate and 233 for media_player. I'll explicitly import all constants for now so checks work,  (likely need noqa on that line, we'll see). I'll provide a pull request for moving the components to use .const after this pull is in.

elupus committed:
Add docstring for climate const

elupus committed:
Add docstring for media_player const

elupus committed:
Explicitly import constants in climate

elupus committed:
Explicitly import constants in media_player

elupus committed:
Add period to climate const

elupus committed:
Add period to media_player const

elupus committed:
Fix some lint errors in climate

elupus committed:
Fix some lint errors in media_player

elupus committed:
Fix lint warnings on climate tests

elupus committed:
Fix lint warnings on group tests

elupus committed:
Fix lint warnings on media_player tests

elupus committed:
Fix lint warnings on state tests

elupus committed:
Adjust indent for state tests
